rfcs/proposed/host_backend_histogram/README.md

```diff
@@ -0,0 +1,73 @@
+
+### parallel_for
+`parallel_for` is an interesting pattern in that it is very generic and embarrassingly parallel. This is close to what we need for `histogram`. However, we cannot simply use it without any added infrastructure. If we were to just use `parallel_for` alone, there would be a race condition between threads when incrementing the values in the output histogram. We should be able to use `parallel_for` as a building block for our implementation, but it requires some way to synchronize and accumulate between threads.
+
+## Proposal
+I believe there are two competing options for `histogram`, which may both have utility in the final implementation depending on the use case.
+
+### Implementation One (Embarrassingly Parallel)
+This method uses temporary storage and a pair of embarrassingly parallel `parallel_for` loops to accomplish the `histogram`.
+1) Determine the number of threads that we will use, perhaps adding some method to do this generically based on the backend.
+2) Create temporary data for the number of threads minus one copy of the histogram output sequence. Thread zero can use the user-provided output data.
+3) Run a `parallel_for` pattern which performs a `histogram` on the input sequence where each thread accumulates into its own copy of the output sequence using the temporary storage to remove any race conditions.
+4) Run a second `parallel_for` over the `histogram` output sequence which accumulates all temporary copies of the histogram into the output histogram sequence. This step is also embarrassingly parallel.
+5) Deallocate temporary storage.
+
+New machinery that will be required here is the ability to query how many threads will be used and also the machinery to check what thread the current execution is using within a brick. Ideally, these can be generic wrappers around the specific backends which would allow a unified implementation for all host backends.
+
+### Implementation Two (Atomics)
+This method uses atomic operations to remove the race conditions during accumulation. With atomic increments of the output histogram data, we can merely run a `parallel_for` pattern.
+
+To deal with atomics appropriately, we have some limitations. We must either use standard library atomics, atomics specific to a backend, or custom atomics specific to a compiler. `C++17` provides `std::atomic<T>`, however, this can only provide atomicity for data which is created with atomics in mind. This means allocating temporary data and then copying it to the output data. `C++20` provides `std::atomic_ref<T>` which would allow us to wrap user-provided output data in an atomic wrapper, but we cannot assume `C++17` for all users. We could look to implement our own `atomic_ref<T>` for C++17, but that would require specialization for individual compilers. OpenMP provides atomic operations, but that is only available for the OpenMP backend.
```

Comment:
> `C++17` provides `std::atomic<T>`, however, this can only provide atomicity for data which is created with atomics in mind. This means allocating temporary data and then copying it to the output data.

I guess you mean that if the output data type is not `std::atomic`, in C++17 there is no standard way to do atomic operations with these data. 

>  We could look to implement our own `atomic_ref<T>` for C++17, but that would require specialization for individual compilers.

I think we do not need a full-fledged `atomic_ref`, we need something like "atomic_cast", and - given that the output data are counters of integral type - that should not be a huge deal.

Reply:
Yes, you are correct. I haven't done a deep look into all compilers in our matrix at this point.  It may be as simple as a reinterpret cast in most cases, but that isn't guaranteed by the standard to work.  I agree this is worth looking at, and may not be so bad.

Reply:
I added a little about my proposed strategy for handling atomics.  I think the best option is to only worry about what we need which is explicitly an atomic increment, and do this using in the best method available.  This allows us the simplest path forward.